docs/libraries/concurrency/actor.hpp/actor3CT3E/index.md

```diff
@@ -0,0 +1,85 @@
+
+However, `actor<T>` is not a panacea. There are several caveats to keep in mind:
+
+1. `thread_local` variables may not retain state from task to task. Given the implementation details of the actor's executor (e.g., it may be scheduled on any number of threads in a thread pool), an actor may jump from thread to thread. Since `thread_local` variables have a per-thread affinity by definition, the variable values may change unexpectedly.
+2. The thread cache penalty paid when an actor changes threads may not be suitable for high-performance/low-latency requirements. There is a cost associated with an actor jumping from one thread to another, and as in the previous case, this may happen depending on the implementation of the executor. If this cache penalty is too expensive for your use case, a dedicated worker thread may be a better fit.
+3. The tasks given to an actor should not block. If the actor must wait for external input (mouse events, network/file IO, etc.) it should be fed in from outside the actor. Because the context of execution is not "owned" by the actor, it cannot presume to block the context waiting for something else to happen, or else it risks hanging (e.g., an unresponsive main thread) or deadlocking (e.g., waiting for a task that cannot complete until this task completes.)
+
+## Example
+
+Say we have a service, `type_rasterizer`, that we'd like to put on a background thread:
+
+```c++
+class image {
+    //...
+};
+
+struct type_rasterizer {
+    void set_text(std::string&& text);
+
+    image rasterize();
+
+    // ... 
+};
+```
+
+In our application, then, we will create an actor that manages an instance of this engine. By giving it the `default_executor`, the actor will run on a thread of the OS-provided thread pool (e.g., GCD on macOS/iOS).
+
+```c++
+struct my_application {
+    stlab::actor<type_rasterizer> _rasterizer(stlab::default_executor,
+                                                "app text rasterizer");
+
+    // ... 
+};
+```
+
+Then as your application is running, you can send "messages" in the form of lambdas to this actor to perform serialized, asynchronous operations. Note the first parameter of the lambda is the `type_rasterizer` itself:
+
+```c++
+void my_application::do_rasterize(std::string&& text) {
+    _rasterizer.send([_text = std::move(text)](type_rasterizer& rasterizer) mutable {
+        // This lambda will execute on the `default_executor`. Note that while in this
+        // lambda, the name of the thread will be the name of the actor. In this case,
+        // "app text rasterizer".
+        rasterizer.set_text(std::move(_text));
+        return rasterizer.rasterize();
+    }).then(stlab::main_executor, [](image my_rasterized_text){
+        draw_image_to_screen(my_rasterized_text);
+    }).detach();
```

Comment:
The detach is worrisome. We should discuss more in slack (and apologies, this week I'm going to be busy getting ready for C++Now so won't be responsive), but most actor systems usually create a "main actor" so you can send a result to the main actor. An actor itself is also a type of "future" so you should be able to `send()` to an actor and not get a future back (I don't know if you want to do that automatically for a `void` result, or make it a separate call). And because of this you should be able to attach an actor to a future as a continuation without generating a new future. Something like: `actor. after(future, [](auto& a, auto future-result) { /* operate on actor type */ });`

I would probably block on `actor` destruction. I detached futures - but that adds a lot of complexity (like pre-exit - and requirement that every continuation can execute unstructured) - but there should be a way to await completion of all calls to an actor so you can have:
```
  actor<type> a;
  //...
  co_await a.complete();
} // actor destructs here.
```

Reply:
I'll have to chat with you more on Slack about this one. I'm thinking they would do well as separate APIs from the ones that do return futures (as I have seen cases where an actor's result will want to get shuttled to another continuation lambda, possibly to another actor, but not required.)